### PR TITLE
Add Withdrawal methods #EP-268

### PR DESCRIPTION
**Pull request by nmdaniil**

### CODE REVIEW

File: XTrophy.cs [hunk 1]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+        public void WithdrawalRequest(Int64 amount, String hashCardNumber)
+        {
+            Require(amount > 0, "Amount must be positive");
+
+            Bytes sender = Info.Sender();
+            WithdrawalData withdrawalData = FrozenBalance.GetOrDefault(sender, new WithdrawalData());
+            Require(withdrawalData.amount == 0, "Withdrawal is already there");
+
+            Int64 balance = Balance.GetOrDefault(sender, 0);
+            Require(balance >= amount, "Not enough XTrophy for Withdrawal");
```
Comment (fomkin) [inline XTrophy.cs:121]: You mean `balance < amount`?

File: XTrophy.cs [hunk 2]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
             } else Error.Throw("XTrophyError: Operation denied");
         }
 
+
+        //// Withdrawal methods
+
+        // Freeze XTrophy.
+        public void WithdrawalRequest(Int64 amount, String hashCardNumber)
```
Comment (fomkin) [inline XTrophy.cs]: `cardNumberHash`

File: XTrophy.cs (hunk 1 above)
Comment (fomkin) [reply XTrophy.cs:121]: `Balance` is 100, `amount` is 50
100 >= 50 is true -> Not enough XTrophy for Withdrawal

File: XTrophy.cs [hunk 3]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+        {
+            Require(amount > 0, "Amount must be positive");
+
+            Bytes sender = Info.Sender();
+            WithdrawalData withdrawalData = FrozenBalance.GetOrDefault(sender, new WithdrawalData());
```
Comment (fomkin) [inline XTrophy.cs]: May it better to use `ContainsKey` + null check to check withdrawal is already there?

File: XTrophy.cs [hunk 4]
```diff
@@ -108,6 +183,12 @@ private void assertIsOwner()
             }
         }
 
+        private void nonEmptyFrozenBalance(Bytes address)
+        {
+            WithdrawalData withdrawalData = FrozenBalance.GetOrDefault(address, new WithdrawalData());
```
Comment (fomkin) [inline XTrophy.cs]: `ContainsKey` and null check.

File: XTrophy.cs [hunk 5]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+            Int64 frozenBalance = FrozenBalance[address].amount;
+            FrozenBalance[address] = new WithdrawalData();
+            Balance[address] = balance + frozenBalance;
+            Log.Event("WithdrawalCanceled", new EventData(address, frozenBalance));
+        }
+
+        // Get withdrawal data.
+        public WithdrawalData GetWithdrawalData(Bytes address)
+        {
+            assertIsOwner();
```
Comment (fomkin) [inline XTrophy.cs]: Are you sure we need this check here? This is public information because it on blockchain.

File: XTrophy.cs [hunk 6]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+
+            return withdrawalData;
+        }
+
+        // Get My withdrawal data.
+        public WithdrawalData MyWithdrawalData()
+        {
+            Bytes sender = Info.Sender();
+            
+            nonEmptyFrozenBalance(sender);
```
Comment (fomkin) [inline XTrophy.cs]: Same as above.

File: XTrophy.cs [hunk 7]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+
+        // Burn frozen XTrophy.
+        public void WithdrawalComplete(Bytes address)
+        {
+            assertIsOwner();
+
+            nonEmptyFrozenBalance(address);
+
+            Int64 frozenBalance = FrozenBalance[address].amount;
+            FrozenBalance[address] = new WithdrawalData();
```
Comment (fomkin) [inline XTrophy.cs]: May be set to null? @Chicker Looks like we forgot to add Remove method to Dictionary.

**Review by fomkin — CHANGES_REQUESTED** [review]
I don't see code that

File: XTrophy.cs (hunk 1 above)
Comment (nmdaniil) [reply XTrophy.cs:121]: This a function Require. 
Require 100 >= 50 else error

File: XTrophy.cs (hunk 3 above)
Comment (nmdaniil) [reply XTrophy.cs]: Я уже пробывал - `ContainsKey` + `null`.
Если установить `null` то будет ошибка `WrongType`

Comment (nmdaniil) [reply XTrophy.cs]: I already tried - `ContainsKey` + `null`
If set to `null` then there will be a `WrongType` error

File: XTrophy.cs (hunk 2 above)
Comment (nmdaniil) [reply XTrophy.cs]: ok

File: XTrophy.cs (hunk 3 above)
Comment (fomkin) [reply XTrophy.cs]: Ok. For future https://github.com/expload/pravda/issues/523

File: XTrophy.cs [hunk 8]
```diff
@@ -92,6 +96,77 @@ public Int64 MyBalance()
...
+            Log.Event("WithdrawalCompleted", new EventData(address, frozenBalance));
```
Comment (fomkin) [inline XTrophy.cs]: A think we should fire `Burn` event too.

**Comment by Chicker** [comment]
We need to add `sender` field to the `Refund` and `Spend` events in order to have info about the company's address (in refund case) and user's address (in spend case)

**Review by Chicker — APPROVED** [review]
Excellent

**Pull request merged**